File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = int(10e15)
    
    time_prev = 0
    min_level, max_level = 1, max(diffs)
    
    while min_level <= max_level:
        level = (min_level + max_level) // 2
        
        spend_time = 0
        for diff, time_cur in zip(diffs, times):
            spend_time += time_cur
            if diff > level:
                spend_time += (time_prev + time_cur) * (diff - level)
                
            time_prev = time_cur
            
            if spend_time >= limit:
                break
            
        if spend_time <= limit:
            answer = min(answer, level)
            max_level = level - 1
        else:
            min_level = level + 1
        
    return answer

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {

    public int solution(int[] diffs, int[] times, long limit) {
        int minLevel = 1, maxLevel = 0;
        for (int d : diffs) maxLevel = Math.max(maxLevel, d);

        int answer = maxLevel;

        while (minLevel <= maxLevel) {
            int level = (minLevel + maxLevel) / 2;

            if (canSolve(diffs, times, limit, level)) {
                answer = level;
                maxLevel = level - 1;
            } else {
                minLevel = level + 1;
            }
        }

        return answer;
    }

    private boolean canSolve(int[] diffs, int[] times, long limit, int level) {
        long spendTime = 0;
        int timePrev = 0;

        for (int i = 0; i < diffs.length; i++) {
            spendTime += times[i];
            if (diffs[i] > level) {
                spendTime += (long)(timePrev + times[i]) * (diffs[i] - level);
            }
            timePrev = times[i];
            if (spendTime > limit) return false;
        }

        return true;
    }
}